### ISSUE TITLE
stage2: Sema panic for comptime array of array of structs

### ISSUE THREAD

**Issue by mitchellh** [ziglang/zig#11139]
```zig
test {
    comptime {
        const S = struct {};

        var cases = [_][1]S{
            [_]S{
                S{},
            },
        };
        _ = cases;
    }
}
```

```
thread 542323 panic: reached unreachable code
Analyzing test2.zig: test2.zig:test.B
      %19 = dbg_stmt(2, 5)
      %20 = extended(struct_decl(anon, Auto, {}, {}, {}) node_offset:21:19
      %21 = alloc_inferred_comptime_mut() node_offset:23:9
      %22 = int(1)
      %23 = array_type(@Ref.one, %20)
      %24 = array_type(%22, %23)
      %25 = coerce_result_ptr(%24, %21)
      %26 = elem_ptr_imm(%25, 0) node_offset:24:17
      %27 = int(1)
      %28 = array_type(%27, %20)
      %29 = coerce_result_ptr(%28, %26)
      %30 = elem_ptr_imm(%29, 0) node_offset:25:18
      %31 = struct_init_empty(%20) node_offset:25:18
    > %32 = store_node(%30, %31) node_offset:25:18
      %33 = validate_array_init_comptime({
        %30 = elem_ptr_imm(%29, 0) node_offset:25:18
      }) node_offset:24:17
      %34 = validate_array_init_comptime({
        %26 = elem_ptr_imm(%25, 0) node_offset:24:17
      }) node_offset:23:28
      %35 = resolve_inferred_alloc(%21) node_offset:23:9
      %36 = load(%21) node_offset:28:13
      %37 = ensure_result_non_error(%36) node_offset:28:13
      %38 = ret_tok(@Ref.void_value) token_offset:30:1
    For full context, use the command
      zig ast-check -t test2.zig


/home/mitchellh/code/go/src/github.com/ziglang/zig/src/value.zig:1087:26: 0x167adeb in value.Value.writeToMemory (zig)
                .Auto => unreachable, // Sema is supposed to have emitted a compile error already
                         ^
/home/mitchellh/code/go/src/github.com/ziglang/zig/src/Sema.zig:17516:22: 0x1642527 in Sema.bitCastVal (zig)
    val.writeToMemory(old_ty, target, buffer);
                     ^
/home/mitchellh/code/go/src/github.com/ziglang/zig/src/Sema.zig:17012:46: 0x152693f in Sema.storePtrVal (zig)
    const bitcasted_val = try sema.bitCastVal(block, src, operand_val, operand_ty, kit.ty, 0);
                                             ^
/home/mitchellh/code/go/src/github.com/ziglang/zig/src/Sema.zig:16982:33: 0x153747b in Sema.storePtr2 (zig)
            try sema.storePtrVal(block, src, ptr_val, operand_val, elem_ty);
                                ^
/home/mitchellh/code/go/src/github.com/ziglang/zig/src/Sema.zig:16935:26: 0x14fbb8f in Sema.storePtr (zig)
    return sema.storePtr2(block, src, ptr, src, uncasted_operand, src, .store);
                         ^
/home/mitchellh/code/go/src/github.com/ziglang/zig/src/Sema.zig:3491:25: 0x13aa7cf in Sema.zirStoreNode (zig)
    return sema.storePtr(block, src, ptr, operand);
                        ^
/home/mitchellh/code/go/src/github.com/ziglang/zig/src/Sema.zig:906:38: 0x125372b in Sema.analyzeBodyInner (zig)
                try sema.zirStoreNode(block, inst);
                                     ^
/home/mitchellh/code/go/src/github.com/ziglang/zig/src/Sema.zig:541:30: 0x124249b in Sema.analyzeBody (zig)
    _ = sema.analyzeBodyInner(block, body) catch |err| switch (err) {
```

**Comment by topolarity** [comment]
Hmm, this failure is on a bitcast from `S` to `[1]S`. The bug is that the array type is popping up here - it should be casting `S` to itself. The good news is that this appears to be fixed by #11128. The 🤷‍♂️ is that I have no idea why

If there's a desire to properly root cause, I'm happy to take a look-see as I work on some changes tomorrow

**Comment by topolarity** [comment]
Think I found the root cause: `Value.elemPtr` will "flatten" elem_ptrs if it notices its parent pointer is an `.elem_ptr`.

For example, an elem_ptr to the third element of an elem_ptr to the second element of X will be flattened to an elem_ptr to the fifth element of X. However, this transformation is only valid if the element type of elem_ptrs is the same. In this bug, an elem_ptr to `S` is being flattened to an elem_ptr of `[1]S` when no flattening should be performed.

This should all be fixed up by #11128
